templates/dhcpd.conf.erb

```diff
@@ -94,6 +94,10 @@ next-server <%= @pxeserver %>;
 <% unless @bootfiles.empty?
      bootf = @bootfiles.sort_by { |arch, file| arch } -%>
 option architecture code 93 = unsigned integer 16 ;
+<% if @ipxe_filename -%>
```

Comment:
Maybe you can build a single array associating all conditions and filenames (by adding multiple arrays):

```puppet
[
  {
    condition => 'exists user-class and option user-class = "iPXE"',
    filename  => 'foo',
  },
  {
    condition => 'option architecture = xxx',
    filename  => 'bar',
  },
  {
    condition => 'option architecture = yyy',
    filename  => 'baz',
  },
]
```

Then you can use a more straightforward template:

```erb
<% data = build_above_array %>
if <%= data.first['condition'] %> {
  filename = "<%= data.first['filename'] %>"
<% data[1..-1].each do |item| %>
} elsif <%= item['condition'] %> {
  filename = "<%= item['filename'] %>"
<% end %>
} else {
  filename "<%= @pxefilename %>";
}
```

Still far from ideal, but avoids the `} els` which is really confusing, and should not cause any trouble if you decide to switch to epp templates at some point (adding arrays is part of the Puppet language).


You can also add a function that formats this array into the proper succession of if / elsif / else, but IMHO, functions are not intended to format strings to feed templates.

Reply:
I like the idea of structured data here. We could enforce it like:
```
Array[Struct[{condition => String[1], filename => String[1]}], 1] $filenames = $dhcp::params::filenames,
```

That should guarantee your code never generates an incorrect structure.